hack/.ci/lib/e2e.sh

```diff
@@ -0,0 +1,253 @@
+#!/bin/bash
+#
+# Copyright (C) 2024 ScyllaDB
+#
+
+set -euExo pipefail
+shopt -s inherit_errexit
+
+source "$( dirname "${BASH_SOURCE[0]}" )/../../lib/kube.sh"
+
+# gather-artifacts is a self sufficient function that collects artifacts without depending on any external objects.
+function gather-artifacts {
+  kubectl create namespace gather-artifacts --dry-run=client -o=yaml | kubectl_apply -f=-
+  kubectl create clusterrolebinding gather-artifacts --clusterrole=cluster-admin --serviceaccount=gather-artifacts:default --dry-run=client -o=yaml | kubectl_apply -f=-
+  kubectl create -n=gather-artifacts pdb must-gather --selector='app=must-gather' --max-unavailable=0 --dry-run=client -o=yaml | kubectl_apply -f=-
+
+  kubectl_create -n=gather-artifacts -f=- <<EOF
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    app: must-gather
+  name: must-gather
+spec:
+  restartPolicy: Never
+  containers:
+  - name: wait-for-artifacts
+    command:
+    - /usr/bin/sleep
+    - infinity
+    image: "${SO_IMAGE}"
+    imagePullPolicy: Always
+    volumeMounts:
+    - name: artifacts
+      mountPath: /tmp/artifacts
+  - name: must-gather
+    args:
+    - must-gather
+    - --all-resources
+    - --loglevel=2
+    - --dest-dir=/tmp/artifacts
+    image: "${SO_IMAGE}"
+    imagePullPolicy: Always
+    volumeMounts:
+    - name: artifacts
+      mountPath: /tmp/artifacts
+  volumes:
+  - name: artifacts
+    emptyDir: {}
+EOF
+  kubectl -n=gather-artifacts wait --for=condition=Ready pod/must-gather
+
+  exit_code="$( wait-for-container-exit-with-logs gather-artifacts must-gather must-gather )"
+
+  kubectl -n=gather-artifacts cp --retries=42 -c=wait-for-artifacts must-gather:/tmp/artifacts "${ARTIFACTS}/must-gather"
+  ls -l "${ARTIFACTS}/must-gather"
+
+  kubectl -n=gather-artifacts delete pod/must-gather --wait=false
+
+  if [[ "${exit_code}" -ne "0" ]]; then
+    echo "Collecting artifacts using must-gather failed"
+    exit "${exit_code}"
+  fi
+}
+
+function gather-artifact-on-exit {
+  gather-artifacts || "Error gathering artifacts" > /dev/stderr
+}
+
+function apply-e2e-workarounds {
+  # Allow admin to use ephemeralcontainers
+  kubectl_create -f=- <<EOF
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: scylladb-e2e:hotfixes
+  labels:
+    rbac.authorization.k8s.io/aggregate-to-admin: "true"
+rules:
+- apiGroups:
+  - ""
+  resources:
+  - pods/ephemeralcontainers
+  verbs:
+  - patch
+EOF
+
+  # FIXME: remove the workaround once https://github.com/scylladb/scylla-operator/issues/749 is done
+  kubectl_create -n=default -f=- <<EOF
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: sysctl
+spec:
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: sysctl
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/name: sysctl
+    spec:
+      containers:
+      - name: sysctl
+        securityContext:
+          privileged: true
+        image: "${SO_IMAGE}"
```

Comment:
validate if SO_IMAGE is non-empty at the beginning of the function

Reply:
I can se the argument both ways, but this result into rechecking it multiple times during the call chain and we do error on undefined variables.

Reply:
done